File: Data Structures & Algorithms/binary-search/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # binary search essentially

        left = 0
        right = len(nums) - 1
        
        while left <= right:
            mid = left + (right - left) // 2
            mid_num = nums[mid]

            if mid_num == target:
                return mid
            elif mid_num < target:
                left = mid + 1
            else:
                right = mid - 1
            
        return -1

File: Data Structures & Algorithms/search-2d-matrix/submission-0.swift
class Solution {
    func searchMatrix(_ matrix: [[Int]], _ target: Int) -> Bool {
        let m = matrix.count
        let n = matrix[0].count
        let count = m * n

        var left = 0
        var right = count - 1
    
        while left <= right{
            let mid = left + (right - left) / 2
            let a = mid / n
            let b = mid % n
            if matrix[a][b] == target{
                return true
            }
            if matrix[a][b] < target{
                left = mid + 1
            }else{
                right = mid - 1
            }
        }
        return false
    }
}
